tred_portable/tred_portable/documentation/man/man1/ntred.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "NTRED 1"
.TH NTRED 1 "2011-08-23" "perl v5.10.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "ntred"
.IX Header "ntred"
ntred \- controller/hub/client interface to a cluster of btred servers
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
To query the servers:
.PP
.Vb 3
\&  ntred [\-m|\-I macro\-file] [\-e code] [\-\-hub hub] [\-\-port port]
\&        [\-N|\-H] [\-T] [\-\-key\-file file] [\-\-filelist file\-list ] [\-L files]
\&        \-\- script\-arguments
.Ve
.PP
To start remote servers and a hub:
.PP
.Vb 9
\&  ntred \-i [\-\-servers server[,server,...]] [\-\-serverlist server\-list]
\&           [\-\-filelist file\-list] [\-\-max\-servers num]
\&           [\-\-allow\-no\-trees] [\-\-no\-secondary\-files]
\&           [\-\-no\-server\-check] [\-\-no\-server\-start] [\-\-no\-hub\-start]
\&           [\-\-old\-dist\-method] [\-\-safe\-mode] [\-\-server\-debug]
\&           [\-m|\-I macro\-file] [\-\-disable\-extensions list] [\-\-enable\-extensions list]
\&           [\-M module] [\-\-btred path\-to\-btred]
\&           [\-\-ssh ssh\-command] [\-\-local]
\&           [\-\-key\-file] [\-\-hub hub] [\-\-port port] [file [...]]
.Ve
.PP
To close all remote servers and a hub:
.PP
.Vb 1
\&  ntred \-\-quit|\-\-ps\-quit
.Ve
.PP
To kill all remote servers and a hub:
.PP
.Vb 1
\&  ntred \-\-kill|\-\-ps\-kill
.Ve
.PP
To manage files on the servers:
.PP
.Vb 1
\&  ntred \-\-list\-files|\-\-list\-changed\-files
\&
\&  ntred \-\-reload\-files [\-\-disable\-extensions list] [\-\-enable\-extensions list]
\&
\&  ntred \-\-reload\-macros [\-m macro\-file]
\&
\&  ntred \-\-load\-files [\-\-filelist file\-list] [file [...]]
\&
\&  ntred \-\-close\-files
\&
\&  ntred \-\-save\-files|\-\-save\-changed\-files [\-s strip\-sfx]
\&       [\-a append\-sfx] [\-p strip\-prefix] [\-r add\-prefix] [\-f out\-fmt]
\&
\&  ntred \-\-dump\-files [\-\-filelist file\-list] [file [...]]
\&
\&  ntred \-\-upload\-file filename < fs\-file
.Ve
.PP
Get help:
.PP
.Vb 3
\&  ntred \-u          for usage (synopsis)
\&  ntred \-h          for help
\&  ntred \-\-man       for the manual page
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This program is able to start one or more btred servers on a set of
host machines over \s-1SSH\s0, create a proxy hub to provide the
communication between the servers and a client, distribute given files
over the servers (provided the servers are able to load the files from
given filenames (eg. they share the files over \s-1NFS\s0), query the servers
using a btred-macro and collect the answers.  It is highly recommended
to use some password-free authentication method for \s-1SSH\s0 (e.g. Kerberos
or \f(CW\*(C`ssh\-agent\*(C'\fR), so that password input is not required each time the
\&\s-1SSH\s0 connection is made (at least two per-host).
.PP
In the client mode, the standard output of the macro is printed to
\&\s-1STDOUT\s0 of the client. \s-1STDERR\s0 is reserved for debugging and information
messages as well as error messages caused by the macros on the
servers. The rest of the error output from a server is stored in a
file <logdir>/ntred\-server\-<host>.log (where <logdir> can be specified
using \f(CW\*(C`\-\-logdir\*(C'\fR, \s-1NTRED_LOGDIR\s0 or \s-1TMP\s0 or \s-1TEMP\s0 environment variable
and defaults to /tmp if none of the previous exists).
.SH "CONFIG FILE"
.IX Header "CONFIG FILE"
Options for ntred can be specified on the command line or in a config
file (default is ~/.ntredrc, but \-\-ntred\-config\-file option can be use
to set a custom config file). The format of this file is simple: each
non-empty line should be of the form:
.PP
.Vb 1
\& option = value
.Ve
.PP
with leading and trailing white-space ignored and with all characters
following a '#' ignored. Option names for the config file are the same
as command-line options, but without leading \- or \-\-.
.SH "OPTIONS"
.IX Header "OPTIONS"
.SS "\s-1QUERY\s0 \s-1MODE\s0 \s-1OPTIONS\s0"
.IX Subsection "QUERY MODE OPTIONS"
.Vb 2
\&  ntred [\-m macro\-file] [\-e code] [\-\-hub hub] [\-\-port port]
\&        [\-N|\-H] [\-T] [\-\-key\-file file] \-\- script\-arguments
.Ve
.IP "\fB\-\-execute|\-e\fR code" 8
.IX Item "--execute|-e code"
This is the query to be executed on the remote servers. It usually
either of a name of the macro defined in the file given as
\&\f(CW\*(C`\-\-macro\-file\*(C'\fR or some Perl one-liner. If omitted, it defaults to
\&'\fIautostart()\fR' and a macro with this name must be defined in
the macro file provided..
.IP "\fB\-\-macro\-file|\-m\fR filename" 8
.IX Item "--macro-file|-m filename"
A file containing macros that are to be sent to the servers and
preloaded on the servers before the particular query code is
evaluated. The query code to be evaluated must be specified in
\&\f(CW\*(C`\-\-execute\*(C'\fR.
.Sp
Note that servers already have the set of macros they obtained on
startup (i.e. \fB\-\-init\fR): it is namely either the default set of
macros (\f(CW\*(C`tred.mac\*(C'\fR) or the set specified via \fB\-\-macro\-file\fR, and beside one of
these also the set specified via \fB\-\-include\-macro\-file\fR.
.IP "\fB\-\-include\-macro\-file|\-I\fR filename" 8
.IX Item "--include-macro-file|-I filename"
A file containing additional set of macros to be sent to the servers.
In query mode, this option behaves just like \f(CW\*(C`\-\-macro\-file\*(C'\fR and is
only provided for compatibility with \f(CW\*(C`btred\*(C'\fR. If both these options
are used, both sets of macros are loaded.
.IP "\fB\-\-hub\fR hostname" 8
.IX Item "--hub hostname"
The hostname of the hub to connect to (defaults to localhost).
.IP "\fB\-\-port\fR port" 8
.IX Item "--port port"
The port on which the hub is listening (defaults to 1500).
.IP "\fB\-\-all\-trees|\-T\fR" 8
.IX Item "--all-trees|-T"
Run the query code on all trees (wrapping the code into a
\&\f(CW\*(C`if ($root) { do {{ CODE }} while TredMacro::NextTree() }\*(C'\fR loop).
.IP "\fB\-\-all\-nodes|\-N\fR" 8
.IX Item "--all-nodes|-N"
Run the query code on all nodes (you still must use \-\-all\-trees or \-T
to process all trees) (wrapping the code into a
\&\f(CW\*(C`while ($this) { CODE ; $this=$this\-\*(C'\fRfollowing }";> loop).
.IP "\fB\-\-all\-nonhidden\-nodes|\-H\fR" 8
.IX Item "--all-nonhidden-nodes|-H"
Run the query code on all nodes that are not hidden (you still must
use \-\-all\-trees or \-T to process all trees). This wraps
the code into a
\&\f(CW\*(C`while ($this) { CODE ; $this=$this\-\*(C'\fRfollowing_visible(\s-1\fIFS\s0()\fR) }";> loop).
.IP "\fB\-\-listed\-files|\-L\fR" 8
.IX Item "--listed-files|-L"
Run the query code only on files specified on the command line
(provided they are already present on some server, i.e. this option
does \fInot\fR make servers load files they don't already
have). Filenames may contain ordinary TrEd suffixes of the form ##tree
or ##tree.node to indicate that the processing should apply only on a
single tree (use in combination with \-N) or a single node (use without
\&\-T and \-N). NTrEd URIs of the form ntred:// are also allowed. Relative
file-names are expanded according to the current working directory
before they are sent to ntred servers.
.IP "\fB\-\-regexp\-files|\-R\fR regular-expression" 8
.IX Item "--regexp-files|-R regular-expression"
Run the query code only on files whose filenames match a given regular
expression.
.IP "\fB\-\-filelist|\-l\fR filename" 8
.IX Item "--filelist|-l filename"
Like \f(CW\*(C`\-\-listed\-files|\-L\*(C'\fR but this time the files to be processed are
listed in the given file rather than on the command line. Both options
may be used together in which case the file-lists are joined.
.IP "\fB\-\-key\-file\fR" 8
.IX Item "--key-file"
This option may be used to provide a file with a session-key which is
necessary for the authentication to the running hub. This defaults to
`~/.ntred_session_key'.
.SS "\s-1HUB\s0 \s-1AND\s0 \s-1SERVER\s0 \s-1MODE\s0 \s-1OPTIONS\s0"
.IX Subsection "HUB AND SERVER MODE OPTIONS"
.Vb 8
\&  ntred \-i [\-\-servers server[,server,...]] [\-\-serverlist server\-list]
\&           [\-\-filelist file\-list] [\-\-max\-servers num]
\&           [\-\-no\-secondary\-files] [\-\-no\-server\-check]
\&           [\-\-no\-server\-start] [\-\-no\-hub\-start] [\-\-old\-dist\-method]
\&           [\-\-safe\-mode] [\-\-server\-debug] [\-\-max\-retries num]
\&           [\-m macro\-file] [\-M module] [\-\-btred path\-to\-btred]
\&           [\-\-ssh ssh\-command] [\-\-local] [\-\-key\-file file]
\&           [\-\-hub hub] [\-\-port port] [file [...]]
.Ve
.IP "\fB\-\-servers\fR list of hosts" 8
.IX Item "--servers list of hosts"
A comma separated list of hosts to run btred servers. The hostname may
be optionally followed by a comma and a port number thus making it
possible to run several btred servers on one host. If the port number
is omitted, it defaults to 1600. See also \f(CW\*(C`\-\-serverlist\*(C'\fR.
.Sp
Ports may be given as a range, e.g. 1600\-1800; in this case btred
server will use the first port from the range that is free.
.Sp
Special syntax can be used to start btred servers over \s-1SGE\s0 cluster
using the task queue (qsub and qdel commands must be in \s-1PATH\s0):
.Sp
.Vb 1
\&  qsub://<job_prefix>:<port_range>
.Ve
.Sp
In this case ntred schedules one btred server task on the \s-1SGE\s0 queue
for each port in the range. It waits at most two seconds times
\&\f(CW\*(C`\-\-max\-retries\*(C'\fR for the first server to start. Tasks of the servers
that are not running by that time are abandoned and deleted from the
queue.
.IP "\fB\-\-serverlist\fR filename" 8
.IX Item "--serverlist filename"
This provides more convenient way to specify servers by providing a
file containing a list of servers, one per line. If neither
\&\f(CW\*(C`\-\-servers\*(C'\fR nor \f(CW\*(C`\-\-serverlist\*(C'\fR is
provided, then the list of servers is read from ~/.ntred_serverlist.
.IP "\fB\-\-filelist\fR filename" 8
.IX Item "--filelist filename"
A list of files to distribute between servers (one filename per line).
Additional files may be given as command-line arguments.
.IP "\fB\-\-no\-secondary\-files\fR" 8
.IX Item "--no-secondary-files"
Don't load \*(L"secondary\*(R" files along with normal files (a file may
require other \- secondary \- file to load along with it; this is
typical for stand-off annotation where one tree is built upon
another).
.IP "\fB\-\-allow\-no\-trees|\-0\fR" 8
.IX Item "--allow-no-trees|-0"
Allow files with no trees (normaly such files are considered broken).
Note: the short flag is \-zero.
.IP "\fB\-\-max\-servers\fR number" 8
.IX Item "--max-servers number"
Limit number of servers to start even if the list of servers contains
more of them.
.IP "\fB\-\-max\-retries\fR number" 8
.IX Item "--max-retries number"
Specifies how many times the hub tries to connect to a btred server
before it gives up.
.IP "\fB\-\-no\-server\-check\fR" 8
.IX Item "--no-server-check"
Skip an initial check for server hosts availability (a dummy attempt
for \s-1SSH\s0 connection).
.IP "\fB\-\-no\-server\-start\fR" 8
.IX Item "--no-server-start"
Don't start new btred servers on the remote hosts. Instead, start a
hub and try to connect to the btred servers already running. This
requires a server-session key to be given on the standard input.
.IP "\fB\-\-no\-hub\-start\fR" 8
.IX Item "--no-hub-start"
Start servers on the remote hosts but don't start a hub. The server
session key required for authentication to the servers will be
printed on the standard output.
.IP "\fB\-\-safe\-mode|\-F\fR" 8
.IX Item "--safe-mode|-F"
Run btred servers in safe mode in which all macros are processed in a
Safe compartment whith some security restrictions.  This mode seems to
be likely to cause btred servers to suffer from memory leaks.
.Sp
In the safe mode, only the following opcodes and opcode-sets are
allowed (see documentation for the Opcode Perl module):
.Sp
.Vb 4
\&  :base_core :base_mem :base_loop :base_math
\&  entereval caller dofile
\&  print entertry leavetry tie untie bless
\&  sprintf localtime gmtime sort require
.Ve
.Sp
plus :base_orig, but the following opcodes (which are forbidden):
.Sp
.Vb 2
\&  getppid getpgrp setpgrp getpriority setpriority pipe_op sselect
\&  select dbmopen dbmclose tie untie
.Ve
.IP "\fB\-\-server\-debug\fR" 8
.IX Item "--server-debug"
Run btred server with \-D flag for some more debugging information.
.IP "\fB\-\-macro\-file|\-m\fR filename" 8
.IX Item "--macro-file|-m filename"
A file containing the default set of macros to be prepared on btred
servers. The file (with exactly the same path) must be visible from
all server hosts.
.IP "\fB\-\-include\-macro\-file|\-I\fR filename" 8
.IX Item "--include-macro-file|-I filename"
A file containing additional set of macros to be prepared on btred
servers. This option is typically used instead of \f(CW\*(C`\-\-macro\-file\*(C'\fR to
load macros from \fBboth\fR \f(CW\*(C`filename\*(C'\fR \fBand\fR the default macro set from
(\f(CW\*(C`tred.mac\*(C'\fR). \f(CW\*(C`\-\-macro\-file\*(C'\fR can still be used in combination with
\&\fB\-\-include\-macro\-file\fR to supply a replacement for \f(CW\*(C`tred.mac\*(C'\fR.
.IP "\fB\-\-enable\-extensions\fR list" 8
.IX Item "--enable-extensions list"
Give a comma-separated list of installed TrEd extension names to
temporarily enable if disabled in the extension configuration.
Use '*' to enable all currently enabled extensions.
.Sp
This option can only be used with \f(CW\*(C`\-\-init\*(C'\fR or \f(CW\*(C`\-\-reload\-macros\*(C'\fR.
.IP "\fB\-\-disable\-extensions\fR list" 8
.IX Item "--disable-extensions list"
Give a comma-separated list of installed TrEd extension names to
temporarily disable if enabled in the extension configuration.
Use '*' to disable all currently enabled extensions.
.Sp
This option can only be used with \f(CW\*(C`\-\-init\*(C'\fR or \f(CW\*(C`\-\-reload\-macros\*(C'\fR.
.IP "\fB\-\-key\-file\fR" 8
.IX Item "--key-file"
Allows to specify a file where the the session-key for a client's
authentication will be stored. It defaults to ~/.ntred_session_key.
.IP "\fB\-\-terminal\-encoding|\-d\fR encoding" 8
.IX Item "--terminal-encoding|-d encoding"
Automatically applies a given character encoding to all stdout output
operations on the servers and command-line arguments.
Only works with Perl >= 5.8.
.IP "\fB\-\-hub\fR hostname" 8
.IX Item "--hub hostname"
The hostname of the local machine the hub will listen on (defaults to
localhost but a machine's hostname may be given to allow remote access
to the hub).
.IP "\fB\-\-port\fR port" 8
.IX Item "--port port"
The port number the hub will be listening on (defaults to 1500).
.IP "\fB\-\-preload\-module|\-M\fR module-name" 8
.IX Item "--preload-module|-M module-name"
This option is passed to the btred command line when starting a btred
server. It makes btred preload a given Perl module at btred startup so
as it is available to all macros (\s-1DOES\s0 \s-1NOT\s0 \s-1WORK\s0 \s-1WITH\s0 \s-1RESTRICTED\s0 \s-1MODE\s0).
This option may be specified more than once with different modules.
.IP "\fB\-\-old\-dist\-method\fR" 8
.IX Item "--old-dist-method"
Use old benchmark-based distribution method.
.SS "\s-1HUB\s0 \s-1CONTROL\s0 \s-1OPTIONS\s0"
.IX Subsection "HUB CONTROL OPTIONS"
.Vb 1
\&  ntred \-\-list\-files|\-\-list\-changed\-files
\&
\&  ntred \-\-reload\-files [\-filelist file\-list] [\-\-listed\-files file [...]]
\&
\&  ntred \-\-reload\-changed\-files
\&
\&  ntred \-\-reload\-macros [\-m macro_file]
\&
\&  ntred \-\-load\-files [\-\-filelist file\-list] [file [...]]
\&
\&  ntred \-\-close\-files
\&
\&  ntred \-\-save\-files|\-\-save\-changed\-files [\-s strip\-sfx]
\&       [\-a append\-sfx] [\-p strip\-prefix] [\-r add\-prefix] [\-f out\-fmt] [\-\-knit]
\&
\&  ntred \-\-quit
\&
\&  ntred \-\-kill|\-\-ps\-kill [\-\-servers server[,server,...]] [\-\-serverlist server\-list]
\&
\&  ntred \-\-break|\-\-ps\-break
\&
\&  ntred \-\-dump\-files [\-\-filelist file\-list] [file [...]]
\&
\&  ntred \-\-upload\-file filename < fs\-file
.Ve
.IP "\fB\-\-init|\-i\fR" 8
.IX Item "--init|-i"
Start remote btred servers and a hub. See \*(L"\s-1HUB\s0 \s-1AND\s0 \s-1SERVER\s0 \s-1MODE\s0 \s-1OPTIONS\s0\*(R".
.IP "\fB\-\-keep\-empty\-servers|\-E\fR" 8
.IX Item "--keep-empty-servers|-E"
Keep alive those servers which do not get any file during the initial
file distribution (by default, unused servers are automatically
closed).
.IP "\fB\-\-quit\fR" 8
.IX Item "--quit"
Sends the hub and all the servers a command to quit.
.IP "\fB\-\-break\fR" 8
.IX Item "--break"
Tells ntred nub to send \s-1USR1\s0 a signal to all running btred servers.
Upon receiving that signal, the servers should stop current processing
and return to the request awaiting state.
.IP "\fB\-\-ps\-kill\fR" 8
.IX Item "--ps-kill"
Similar to \f(CW\*(C`\-\-break\*(C'\fR but tries to identify btred server processes by
looking at the output of the system command \f(CW\*(C`ps x \-o pid,command\*(C'\fR. This may help if
\&\f(CW\*(C`killall \-USR1 btred\*(C'\fR doesn't work.
.IP "\fB\-\-kill\fR" 8
.IX Item "--kill"
Runs \f(CW\*(C`killall \-9 ntred\*(C'\fR on the local machine and and \f(CW\*(C`killall \-9 btred\*(C'\fR on
the server hosts listed with \-\-servers, \-\-serverlist, or in
~/.ntred_serverlist.
.IP "\fB\-\-ps\-kill\fR" 8
.IX Item "--ps-kill"
Similar to \f(CW\*(C`\-\-kill\*(C'\fR but tries to identify btred server processes by
looking at the output of the system command \f(CW\*(C`ps x \-o pid,command\*(C'\fR. This may help if
\&\f(CW\*(C`killall \-TERM btred\*(C'\fR doesn't work.
.IP "\fB\-\-list\-files\fR" 8
.IX Item "--list-files"
List all files currently open on servers.
.IP "\fB\-\-list\-changed\-files\fR" 8
.IX Item "--list-changed-files"
List all files that have been changed by some macro. Note, that a macro
has to claim that the file was changed by setting
\&\f(CW$TredMacro::FileChanged\fR variable to 1, otherwise the btred server
would never notice.
.IP "\fB\-\-listed\-files|\-L\fR" 8
.IX Item "--listed-files|-L"
Apply request only on listed files (currently only works for queries
and \f(CW\*(C`\-\-reload\-files\*(C'\fR request).
.IP "\fB\-\-count\-files\fR" 8
.IX Item "--count-files"
Query number of files open on each server.
.IP "\fB\-\-reload\-files\fR" 8
.IX Item "--reload-files"
Send a command to the btred servers to reload all open files.  If
\&\f(CW\*(C`\-\-filelist\*(C'\fR or \f(CW\*(C`\-\-listed\-files\*(C'\fR options are given, reload only
files occuring in the given lists (all other files remain intact in
servers' memory).
.IP "\fB\-\-reload\-changed\-files\fR" 8
.IX Item "--reload-changed-files"
Send a command to the btred servers to reload files that have been
modified since (re)loaded.
.IP "\fB\-\-reload\-macros\fR" 8
.IX Item "--reload-macros"
Send command to the btred servers to reload the initial macro file.
If \f(CW\*(C`\-m\*(C'\fR (\f(CW\*(C`\-\-macro\-file\*(C'\fR) is specified, the servers use the given
macro-file instead of the original one (specified when initializing
btred servers). Note, that the file (with exactly the same path) must
be visible from all server hosts.
.Sp
Options \f(CW\*(C`\-\-disable\-extensions\*(C'\fR and \f(CW\*(C`\-\-enable\-extensions\*(C'\fR can be used
to define a set of extensions to use.
.IP "\fB\-\-load\-files\fR" 8
.IX Item "--load-files"
Send a command to the hub to distribute files given on the
command-line or those specified using \f(CW\*(C`\-\-filelist\*(C'\fR of in
~/.ntred_filelist to the servers.
.Sp
Warning: Do not load files already in servers' memory, otherwise you
may get duplicates (depending on to which server the files get
distributed). Use \f(CW\*(C`\-\-list\-files\*(C'\fR to see which files are already loaded.
.Sp
See also \f(CW\*(C`\-\-reload\-files\*(C'\fR.
.IP "\fB\-\-close\-files\fR" 8
.IX Item "--close-files"
Send a command to the btred servers to close all open files.
.IP "\fB\-\-save\-files\fR" 8
.IX Item "--save-files"
Send a command to the servers to save all open files.  The filenames
of the saved files may be modified using
\&\f(CW\*(C`\-\-add\-prefix\*(C'\fR, \f(CW\*(C`\-\-strip\-prefix\*(C'\fR,
\&\f(CW\*(C`\-\-strip\-suffix\*(C'\fR,
\&\f(CW\*(C`\-\-append\-suffix\*(C'\fR.
.IP "\fB\-\-save\-changed\-files\fR" 8
.IX Item "--save-changed-files"
Same as \f(CW\*(C`\-\-save\-files\*(C'\fR except that only files that have been changed
by some macro will be saved. Note, that a macro has to claim that the
file was changed by setting \f(CW$TredMacro::FileChanged\fR variable to 1,
otherwise the btred server would never notice. See also
\&\f(CW\*(C`\-\-list\-changed\-files\*(C'\fR.
.IP "\fB\-\-knit|K\fR ALL|NONE|name1,name2,..." 8
.IX Item "--knit|K ALL|NONE|name1,name2,..."
If a file is saved, save/update also listed types of reffiles the file
pulled data from. For the moment, this only makes sense with the
\&\f(CW\*(C`PML\*(C'\fR backend which supports so called \fBknitting\fR, i.e. a method to
pull certain data from external resources and push it back (with all
changes) to the original position in the resource when saving the
file. This option allows to list the types of resources (in \s-1PML\s0 the
types are the reference names listed in the \s-1PML\s0 schema) which should
be saved. Default is \f(CW\*(C`NONE\*(C'\fR. This type of resources doesn't include
so called secondary files.
.IP "\fB\-\-dump\-files\fR" 8
.IX Item "--dump-files"
Dumps given files or trees to standard output in \s-1FS\s0 format as they are
in memory of the btred servers. Individual dumps are separated with
`//FSEND' preceded and followed by two newline characters.  To output
a single tree, follow the filename with \f(CW\*(C`##n\*(C'\fR suffix where \f(CW\*(C`n\*(C'\fR is
the absolute position of the tree in the file (starting from one).
The following example shows how \f(CW\*(C`csplit\*(C'\fR command can be used to
save individual dumps into separate files:
.Sp
.Vb 1
\&  ntred \-\-dump <files> | csplit \-z \-f out \-b \*(Aq%d.fs\*(Aq \- \*(Aq/\e/\e/FSEND/+2\*(Aq \*(Aq{*}\*(Aq
.Ve
.Sp
To merge these separate files into one huge \s-1FS\s0 file, use
.Sp
.Vb 1
\&  any2any \-m hugeout.fs out*.fs
.Ve
.IP "\fB\-\-upload\-file\fR" 8
.IX Item "--upload-file"
This command is kind of a reverse to dump. It takes a filename from
the first command line argument and a \s-1FS\s0 file from the standard input
and sends them to the btred servers. The server possessing the file
with the given filename replaces its own in-memory copy of the file
with the one provided on the standard input. To update a single tree,

tred_portable/tred_portable/dependencies/lib/perl5/Treex/PML/Backend/NTRED.pm
package Treex::PML::Backend::NTRED;
use Treex::PML;
use Storable qw(nfreeze thaw);
use MIME::Base64;
use Treex::PML::IO;
use strict;

use vars qw($VERSION);
BEGIN {
  $VERSION='2.07'; # version template
}

use vars qw($ntred);


$ntred='ntred';

sub test {
  my ($filename,$encoding)=@_;
  return $filename=~m(^ntred://);
}

sub open_backend {
  my ($filename, $mode, $encoding)=@_;
  my $fh = undef;
  my $cmd = "";
  return unless $filename=~m(^ntred://(.*)$);
  $filename=$1;
  $filename=~s/@/##/;
  if ($filename) {
    if ($mode eq 'w') {
      open($fh,'|-',$ntred, '-Q', '--upload-file', $filename) || die "Failed to start NTrEd Client '$ntred': $!\n";
    } else {
      open($fh,'-|',$ntred, '-Q', '--dump-files', $filename) || die "Failed NTrEd Client '$ntred': $!\n";
    }
  }
  return Treex::PML::IO::set_encoding($fh,$encoding);
}

sub close_backend {
  my ($fh)=@_;
  return $fh && $fh->close();
}

sub read {
  my ($fd,$fs)=@_;
  my $data = do{{ local $/; <$fd>}};
  unless (defined $data and length $data) {
    die "NTrEd Client returned no data\n";
  }
  my $fs_files = Storable::thaw(decode_base64($data));
  undef $data;
  my $restore = $fs_files->[0];
  if (ref($restore)) {
    my $api_version = $restore->[6];
    if ($api_version ne $Treex::PML::API_VERSION) {
      warn "Warning: the binary content obtained via Treex::PML::Backend::NTRED from ".$fs->filename." is a dump of structures created by possibly incompatible Treex::PML API version $api_version (the current Treex::PML API version is $Treex::PML::API_VERSION)\n";
    }
    $fs->changeFS($restore->[0]);
    $fs->changeTrees(@{$restore->[1]});
    $fs->changeTail(@{$restore->[2]});
    $fs->[13]=$restore->[3]; # metaData
    $fs->changePatterns(@{$restore->[4]});
    $fs->changeHint($restore->[5]);
    $fs->FS->renew_specials();
  }
}



sub write {
  my ($fh,$fsfile)=@_;
  my $dump= [$fsfile->FS,
	     $fsfile->treeList,
	     [$fsfile->tail],
	     $fsfile->[13], # metaData
	     [$fsfile->patterns],
	     $fsfile->hint,
	     $Treex::PML::API_VERSION
	     # CAUTION: when adding to this list, don't forget to do the same
	     # in btred DUMP request handler
	    ];
  eval {
    print $fh (encode_base64(Storable::nfreeze([$dump])));
    print $fh ("\n");
  };
}

1;
__END__

=pod

=head1 NAME

Treex::PML::Backend::NTRED - Treex::PML I/O backend for exchanging data with remote ntred servers.

=head1 SYNOPSIS

use Treex::PML;
Treex::PML::AddBackends(qw(NTRED))

my $document = Treex::PML::Factory->createDocumentFromFile('ntred:///some/file');
...
$document->save();

my $doc_frag = Treex::PML::Factory->createDocumentFromFile('ntred:///some/file@10');
...
$doc_frag->save();

=head1 DESCRIPTION

This module implements a Treex::PML input/output backend which
exchanges data with remote ntred servers. It uses the external
C<ntred> command-line client to communicate to the servers.

The backend accepts any document whose URL is of the form

  ntred:///some/file

or

  ntred:///some/file@N

where C<N> is an integer. In first form can be used to retrieve/save
back a document whose local name C</some/file> from the in memory of
an ntred server.  The later form can be used to retrieve a partial
document containing only the Nth tree (the index is 0-based) of the
specified document. When this partial document is saved back, the Nth
tree in the in-memory representation of the document on the server is
updated, leaving other trees intact.

=head2 SEE ALSO

TrEd toolkit (L<http://ufal.mff.cuni.cz/~pajas/tred>)

=head2 REFERENCE

=over 4

=item $Treex::PML::Backend::NTRED::ntred

This variable may be used to set-up the path to 'ntred' client program.

=back

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006-2010 by Petr Pajas

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.2 or,
at your option, any later version of Perl 5 you may have available.

=cut
